Leanplum/Leanplum-iOS-SDK
# [SDK-24] Defer IAM

What              | Where/Who
------------------|----------------------------------------
JIRA Issue        | [SDK-24](https://leanplum.atlassian.net/browse/SDK-24)

## Background
Do not show In-app messages on particular screens. Defer the display of the message until a suitable screen is shown. 

This should work for both built-in and custom in-app message templates.

## Implementation
On Action trigger, check if the action should be deferred. If yes, persist the context and run it on the next suitable controller using the swizzling of `viewDidAppear`.

If there is more than one deferred message, the messages will be executed all one after another - triggered once the previous one is dismissed. This is done by swizzling the `dismissViewControllerAnimated` method.
See `LPDeferMessageManager` for details.

Cases where the deferred message will not be presented - for example, Push Pre-Permission where the permission has already been accepted, are also handled. In that case, `triggerAction` handledBlock is false.

The logic also defers messages on the UIAlertController so the message is not shown inside the alert.

## Testing steps
Set the deferred controllers:
`[Leanplum deferMessagesForViewControllers:@[LPViewController.class]];`
`Leanplum.deferMessages([HomeViewController.self])`

Trigger an In-app message on that controller. The message should not be displayed. 
Navigate to another controller. The message should be displayed automatically.

## Is this change backwards-compatible?
